scalding-commons-macros/src/main/scala/com/twitter/scalding/commons/macros/impl/ordser/UnionLike.scala

```diff
@@ -0,0 +1,279 @@
+        }.getOrElse[Tree](q"0")
+
+        existing match {
+          case Some(t) =>
+            Some(q"""
+              if($valueA == $idx) {
+                $commonCmp
+              } else {
+                $t
+              }
+            """)
+          case None =>
+            Some(q"""
+                if($valueA == $idx) {
+                  $commonCmp
+                } else {
+                  sys.error("Unable to compare unknown type")
+                }""")
+        }
+    }.get
+
+    q"""
+        val $valueA: Int = $inputStreamA.readByte.toInt
+        val $valueB: Int = $inputStreamB.readByte.toInt
+        val $idxCmp: Int = $valueA.compare($valueB)
+        if($idxCmp != 0) {
+          $idxCmp
+        } else {
+          $compareSameTypes
+        }
+      """
+  }
+
+  def put(c: Context)(inputStream: c.TermName, element: c.TermName)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val innerArg = freshT("innerArg")
+    subData.foldLeft(Option.empty[Tree]) {
+      case (optiExisting, (idx, tpe, optiTBuf)) =>
+        val commonPut: Tree = optiTBuf.map { tBuf =>
+          q"""val $innerArg: $tpe = $element.asInstanceOf[$tpe]
+              ${tBuf.put(inputStream, innerArg)}
+              """
+        }.getOrElse[Tree](q"()")
+
+        optiExisting match {
+          case Some(s) =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $inputStream.writeByte($idx.toByte)
+              $commonPut
+            } else {
+              $s
+            }
+            """)
+          case None =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $inputStream.writeByte($idx.toByte)
+              $commonPut
+            }
+            """)
+        }
+    }.get
+  }
+
+  def length(c: Context)(element: c.Tree)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): LengthTypes[c.type] = {
+    import c.universe._
+
+    val prevSizeData = subData.foldLeft(Option.empty[Tree]) {
+      case (optiTree, (idx, tpe, tBufOpt)) =>
+
+        val baseLenT: Tree = tBufOpt.map{ tBuf =>
+          tBuf.length(q"$element.asInstanceOf[$tpe]") match {
+            case m: MaybeLengthCalculation[_] =>
+              m.asInstanceOf[MaybeLengthCalculation[c.type]].t
+
+            case f: FastLengthCalculation[_] =>
+              q"""Some(Right(${f.asInstanceOf[FastLengthCalculation[c.type]].t})) : Option[Either[Int, Int]]"""
+
+            case _: NoLengthCalculationAvailable[_] =>
+              return NoLengthCalculationAvailable(c)
+            case e => sys.error("unexpected input to union length code of " + e)
+          }
+        }.getOrElse(q"Some[Either[Int, Int]](Right(1))")
+
+        val lenT = q"""
+        $baseLenT.map{ either =>
+        either match {
+                case(Left(l)) => Right(l + 1) : Either[Int, Int]
```

Comment:
It's easy to lose track of what these types mean (which is the Left, Right)?

Can you introduce types for these?